java-tools/OsmAndMapCreatorUtilities/src/main/java/net/osmand/obf/preparation/IndexRouteCreator.java

```diff
@@ -623,27 +628,39 @@ private void indexHighwayRestrictions(Entity e, OsmDbAccessorContext ctx) throws
 					type = MapRenderingTypes.RESTRICTION_ONLY_STRAIGHT_ON;
 				}
 				if (type != -1) {
-					ctx.loadEntityRelation((Relation) e);
-					Collection<RelationMember> fromL = ((Relation) e).getMembers("from"); //$NON-NLS-1$
-					Collection<RelationMember> toL = ((Relation) e).getMembers("to"); //$NON-NLS-1$
-					Collection<RelationMember> viaL = ((Relation) e).getMembers("via"); //$NON-NLS-1$
-					if (!fromL.isEmpty() && !toL.isEmpty()) {
-						RelationMember from = fromL.iterator().next();
-						RelationMember to = toL.iterator().next();
-						if (from.getEntityId().getType() == EntityType.WAY) {
-							if (!highwayRestrictions.containsKey(from.getEntityId().getId())) {
-								highwayRestrictions.put(from.getEntityId().getId(), new ArrayList<>());
-							}
-							RestrictionInfo rd = new RestrictionInfo();
-							rd.toWay = to.getEntityId().getId();
-							rd.type = type;
-							if(!viaL.isEmpty()) {
-								RelationMember via = viaL.iterator().next();
-								if(via.getEntityId().getType() == EntityType.WAY) {
-									rd.viaWay = via.getEntityId().getId();
+					ctx.loadEntityRelation(r);
+					Collection<RelationMember> fromL = r.getMembers("from"); //$NON-NLS-1$
+					Collection<RelationMember> toL = r.getMembers("to"); //$NON-NLS-1$
+					Collection<RelationMember> viaL = r.getMembers("via"); //$NON-NLS-1$
+					if (!toL.isEmpty()) {
+						for (RelationMember from : fromL) {
+							if (from.getEntityId().getType() == EntityType.WAY) {
+								if (!highwayRestrictions.containsKey(from.getEntityId().getId())) {
+									highwayRestrictions.put(from.getEntityId().getId(), new ArrayList<>());
+								}
+
+								List<RestrictionInfo> rdList = highwayRestrictions.get(from.getEntityId().getId());
+								for (RelationMember to : toL) {
+									RestrictionInfo rd = new RestrictionInfo();
```

Comment:
This code needs to be simplififed, used to be 5 inners ifs, but now 4 ifs + 2 fors too complex

Reply:
Thanks for your feedback. I totally agree.
I guess this would be a good use case for guard clauses to return immediately when one of the member lists is empty. This would eliminate two nesting levels. Additionally, the checks for `val` (currently line 592) and `type` (currently line 630) could also be written as guard clauses, further reducing the nesting of the whole function.
Would this approach be okay for you? If so, I would add a commit containing only these refactorings.